packages/snap/src/index.ts

```diff
@@ -1,33 +1,92 @@
-import { OnRpcRequestHandler } from '@metamask/snaps-types';
-import { panel, text } from '@metamask/snaps-ui';
+import { providerErrors, rpcErrors } from '@metamask/rpc-errors';
+import { DialogType, OnRpcRequestHandler } from '@metamask/snaps-types';
+import { copyable, heading, panel, text } from '@metamask/snaps-ui';
+import { SLIP10Node } from '@metamask/key-tree';
+import {
+  add0x,
+  assert,
+  bytesToHex,
+  remove0x,
+  valueToBytes,
+} from '@metamask/utils';
+import { sign as signEd25519 } from '@noble/ed25519';
+import { sign as signSecp256k1 } from '@noble/secp256k1';
+
+import type { GetBip32PublicKeyParams, SignMessageParams } from './types';
 
 /**
  * Handle incoming JSON-RPC requests, sent through `wallet_invokeSnap`.
  *
  * @param args - The request handler args as object.
- * @param args.origin - The origin of the request, e.g., the website that
- * invoked the snap.
  * @param args.request - A validated JSON-RPC request object.
- * @returns The result of `snap_dialog`.
+ * @returns The result of the method.
  * @throws If the request method is not valid for this snap.
  */
-export const onRpcRequest: OnRpcRequestHandler = ({ origin, request }) => {
+export const onRpcRequest: OnRpcRequestHandler = async ({ request }) => {
   switch (request.method) {
-    case 'hello':
-      return snap.request({
+    case 'getPublicKey':
+      return await snap.request({
+        method: 'snap_getBip32PublicKey',
+        params: request.params as unknown as GetBip32PublicKeyParams,
+      });
+
+    case 'signMessage': {
+      const { message, curve, ...params } = request.params as SignMessageParams;
+      const json = await snap.request({
+        method: 'snap_getBip32Entropy',
+        params: {
+          ...params,
+          curve,
+        },
+      });
+      const node = await SLIP10Node.fromJSON(json);
+
+      assert(node.privateKey);
+      assert(curve === 'ed25519' || curve === 'secp256k1');
+
+      const approved = await snap.request({
         method: 'snap_dialog',
         params: {
-          type: 'confirmation',
+          type: DialogType.Confirmation,
           content: panel([
-            text(`Hello, **${origin}**!`),
-            text('This custom confirmation is just for display purposes.'),
+            heading('Signature request'),
             text(
-              'But you can edit the snap source code to make it do something, if you want to!',
+              `Do you want to ${curve} sign "${message}" with the following public key?`,
             ),
+            copyable(add0x(node.publicKey)),
           ]),
         },
       });
-    default:
-      throw new Error('Method not found.');
+
+      if (!approved) {
+        throw providerErrors.userRejectedRequest();
+      }
+
+      if (curve === 'ed25519') {
```

Comment:
nit: these two signing blocks are identical except for the `sign` function. Maybe we can pull out `valueToBytes` and `remove0x` etc.